### PR TITLE
proxy: Modify registry auth url for proxy cache validation (PROJQUAY-4585)

### PR DESCRIPTION
**Pull request by harishsurf**
When authenticating against a registry, if the `www-authenticate` header doesn't specify a `service` parameter, it should be excluded from the registry auth url.

Eg:
```
$  curl -I -L https://registry.ci.openshift.org/v2/
HTTP/2 401
content-type: application/json; charset=utf-8
docker-distribution-api-version: registry/2.0
www-authenticate: Bearer realm="https://registry.ci.openshift.org/openshift/token"
```

the `www-authenticate` header doesn't specify a `service` parameter. Prior to this change, the malformed url request failed with a `401`
```
$ curl https://registry.ci.openshift.org/openshift/token\?service\=None\&scope\=repository:something-totally-fake:pull
{"details":"repository name \"something-totally-fake\" invalid: it must be of the format \u003cproject\u003e/\u003cname\u003e"} 
```

Signed-off-by: harishsurf <hgovinda@redhat.com>

### CODE REVIEW

**Comment by ibazulic** [comment]
I am not sure we should allow this. The [authentication specification](https://docs.docker.com/registry/spec/auth/token/) that Docker and all registries indicates that the field is mandatory.

According to the [RFC 6750](https://www.rfc-editor.org/rfc/rfc6750#page-7), which defines the Oauth protocol, the parameter `realm` is apparently optional, but the [RFC-2617](https://www.rfc-editor.org/rfc/rfc2617) defines it as mandatory. Specifically:

> The authentication parameter realm is defined for all authentication
>    schemes:
> 
>       realm       = "realm" "=" realm-value
>       realm-value = quoted-string
> 
>    The realm directive (case-insensitive) is required for all
>    authentication schemes that issue a challenge. The realm value
>    (case-sensitive), in combination with the canonical root URL (the
>    absoluteURI for the server whose abs_path is empty; see section 5.1.[2](https://www.rfc-editor.org/rfc/rfc2617#ref-2)
>    of [2]) of the server being accessed, defines the protection space.
>    These realms allow the protected resources on a server to be
>    partitioned into a set of protection spaces, each with its own
>    authentication scheme and/or authorization database. The realm value
>    is a string, generally assigned by the origin server, which may have
>    additional semantics specific to the authentication scheme. Note that
>    there may be multiple challenges with the same auth-scheme but
>    different realms.

In other words, the `realm` or `service` **must be** returned by the authentication endpoint, after which we **may** include it in our own challenge.

**Comment by harishsurf** [comment]
@ibazulic , are you referring to the `realm` parameter? This change doesn't remove this parameter. It excludes `service` parameter from the url if the `www-authenticate` header doesn't specify it. 
I couldn't find any related material for `service` usage in the RFC doc. I did check with @dmage whether this is a bug on the openshift token endpoint, may be he can say more..

**Comment by ibazulic** [comment]
Yeah, but in this case I think that the `service` parameter plays the role of the `realm`. If my logic is wrong here, then this discussion doesn't matter, of course :-)

**Review by flavianmissi — DISMISSED** [review]
@ibazulic the realm is sent regardless of the service being present or not, the condition is whether to use the service (+scope) along with the realm, or to just use the realm when no service is present.

@harishsurf I see that the scope is also being removed when no service is sent - have you tested this on a few registries? An easy way to test it is to configure a proxy org. Before creating a proxy cache config the backend will try to get an auth token from the registry and return an error if it fails.
I think we should also notify QE to do a full regression test after this one goes in, just to be on the safe side.

**Review by dmage — CHANGES_REQUESTED** [review]
Lack of service doesn't mean that we need to reject scope. Let's not mix them.

**Review by flavianmissi — APPROVED** [review]
LGTM
